### ISSUE TITLE
BUG: astype(object) downcasts for datetime-dtype

### ISSUE THREAD

**Issue by h-vetinari** [numpy/numpy#12550]
The dtype `np.object_` is the most general catch-all type to contain arbitrary python objects. As such, `.astype(object)` should preferably not upcast, but certainly never *down*cast:

### Reproducing code example:

```python
>>> import numpy as np
>>> arr = np.array([10 ** 18], dtype='M8[ns]')
>>> arr
array(['2001-09-09T01:46:40.000000000'], dtype='datetime64[ns]')
>>> arr[0]
numpy.datetime64('2001-09-09T01:46:40.000000000')
>>> arr.astype(object)
array([1000000000000000000], dtype=object)
>>> arr.astype(object)[0]
1000000000000000000
```

The expected outcome for the last two lines is:
```
>>> arr.astype(object)
array([numpy.datetime64('2001-09-09T01:46:40.000000000')], dtype=object)
>>> arr.astype(object)[0]
numpy.datetime64('2001-09-09T01:46:40.000000000')
```
as well as
```
>>> arr.astype(object)[0] == arr[0]
True
```

### Numpy/Python version information:

```
numpy: 1.15.4
python: 3.7.1
```

**Comment by eric-wieser** [comment]
Note that the result is usually a `datetime.timedelta`, if one is possible to create:
```
>>> np.array([10 ** 10], dtype='M8[us]').astype(object)
array([datetime.datetime(1970, 1, 1, 2, 46, 40)], dtype=object)
```

**Comment by h-vetinari** [comment]
> Note that the result is usually a datetime.~timedelta~datetime, if one is possible to create:

Fair enough, as long as it's not an int... ;-)

**Comment by eric-wieser** [comment]
Well in this case, it doesn't fit in a date time, so we don't have that option.

Is it better to return:
1. Only datetime64 objects
2. Only rounded / truncated date times
3. A mixture of datetime and int
4. A mixture of datetime64 and datetime

Right now, we do 3.

**Comment by h-vetinari** [comment]
To me the answer is clearly 1 (should be clear that a mixture is bound to inconsistencies...).

TBH, I was really surprised that (from your example):
```
>>> np.array([10 ** 10], dtype='M8[us]').astype(object)
array([datetime.datetime(1970, 1, 1, 2, 46, 40)], dtype=object)
```
rather than
```
>>> np.array([10 ** 10], dtype='M8[us]').astype(object)
array([numpy.datetime64('1970-01-01T02:46:40.000000')], dtype=object)
```

This is also inconsistent because passing an `np.datetime64` to a constructor with `dtype=object` does not change it to `datetime.datetime`, so I was really surprised that `.astype(object)` does something different.

```
>>> np.array([np.datetime64('1970-01-01T02:46:40.000000')], dtype=object)
array([numpy.datetime64('1970-01-01T02:46:40.000000')], dtype=object)
>>>
>>> np.array([np.datetime64('1970-01-01T02:46:40.000000')]).astype(object)
array([datetime.datetime(1970, 1, 1, 2, 46, 40)], dtype=object)
```

**Comment by seberg** [comment]
Well, the current state does seem strange. But I am not sure I like to change the behaviour so something that on first sight seems rather hypothetical? (I mean changing the behaviour to returning datetime64, which seems pretty different.)

Do you actually work with dates past 10000 years? If that is a real use case, maybe we should see that python allows it rather?

**Comment by h-vetinari** [comment]
@seberg 
> Well, the current state does seem strange. But I am not sure I like to change the behaviour because of something that on first sight seems rather hypothetical? (I mean changing the behaviour to returning datetime64, which seems pretty different.)

I don't really get that point. The underlying values were `datetime64` to begin with, why should `astype(object)` change the values (rather than just the dtype of the array)?
I understand of course that other `astype`-cases deliberately change the values (indeed, that's often the whole point of the method call), but `object` is special as the most general container, and should never need to cast at all.

> Do you actually work with dates past 10000 years? If that is a real use case, maybe we should see that python allows it rather?

This is not an issue of date range limitations (to me). I'm writing lots of parametrized tests for a larger pandas-PR, and this discrepancy between `dtype=object` in the constructor and `.astype(object)` caught me out.

**Comment by seberg** [comment]
This is annoying, but... If you look closer, you will notice that also other types cast to the python version. For datetime this is a bigger step/change and also annoying.

What I mean is that we use `.item()`'s "convert to a python object" logic for datetime:
     * This assumes that numpy datetimes can be safely cast to python datetimes (which is not quite true)
     * **But:** Changing this is a pretty large change that needs to be thoroughly discussed (at least as far as I can see). For example, what will happen with matplotlib, and all the other users down there that (unwittingly) are using it to get datetime objects.

So, yes, this is not a safe cast. And yes it is utterly broken. But, I need a a lot more to be convinced that changing the cast from returning `datetime` objects to returning numpy `datetime64` objects will not have a huge impact on downstream.

If such a fix would actually fix bugs for downstream in the long run, it would be more compelling. But it seems to me that for most users the bug will be very mild...

**Comment by h-vetinari** [comment]
@seberg 
We got side-tracked a bit with the `datetime` vs `datetime64` - I see your points, even though I believe downstream effort is not a good reason to keep something in an "utterly broken" state. That's why there's version numbers. ;-)

I can work around the `datetime`/`datetime64` thing, but the `ints` that get returned in the OP are really a killer.

**Comment by seberg** [comment]
I noticed that years ago I once glanced at changing the default behaviour for all types here (so that the object cast would retain the numpy type). But I am not quite sure we should do it.

**Comment by eric-wieser** [comment]
.astype(np.generic) might be an interesting way to allow spelling that,
@seberg

**Comment by seberg** [comment]
There seem to be some similar odd casts noted in gh-5180, will close that one in favor if this one.

**Comment by eric-wieser** [comment]
This is a duplicate of #8546

**Comment by gerritholl** [comment]
> Is it better to return:
> 
>     1. Only datetime64
> 
>     2. Only rounded / truncated datetime
> 
>     3. A mixture of datetime and int
> 
>     4. A mixture of datetime64 and datetime
> 
> Right now, we do 3.

How about adding a keyword argument to  `.astype` and related methods, such that the user can choose between those four options?  I usually want a truncated datetime, but maybe sometimes a datetime64.  I never want int.

**Comment by eraoul** [comment]
There seems to be tons of discussion, but it's clear that when I do x.astype(datetime) and it returns an integer instead of a datetime is a serious bug. Just wasted far too long tracking this down, and was completely shocked to realize that the type was wrong after ".astype(datetime)". Error message or options would be great.

**Comment by seberg** [comment]
@eraoul yes, I think we should do something. An error, or at least a warning seems fair to me (it is within the casting machinery, so it may take a bit of care). PRs are welcome, I can give you pointers if you want to look into it.

**Comment by h-vetinari** [comment]
I commented a bit on the sister issue #7619, suggesting that `.item` could also just return `np.datetimes64` instead of trying (and failing) to cast to `datetime.datetime`, whereas @eric-wieser floated the possibility of separating the (currently shared/similar) semantics of `.astype(object)` from those of `.item`.

I'm fine with either solution, but certainly, the `astype`-stuff is more pressing IMO.

**Comment by seberg** [comment]
It seems we converged to the following when talking about it today:

1. We will deprecate any return which is not a python datetime
2. It will give an error in the future.
3. If someone has a use-case, we may try to provide a utility function.

Note that this would include the conversion of `NaT` to `None`. Do you think that this will help with the issue? Is the deprecation of conversion to `None` reasonable?

Please just comment on this, doing the actual change should be fairly straight forward.

**Comment by h-vetinari** [comment]
@seberg 
Thanks for tackling this. I dislike that `.astype(object)` will still cast `np.datetime64` to python datetime _at all_, but I'm realize that providing an consistent and realistic path for migration is a challenge.

Still, I think that valid `np.datetime64` that are "just" cast to object shouldn't raise an error, just because it's outside of what `datetime.datetime` can handle. The output there is already demonstrably wrong, so why not return `np.datetime64` for those cases, and deprecate returning `datetime.datetime`?

**Comment by seberg** [comment]
@h-vetinari the thing is, if we stick with current datetimes, you would get a return with inhomogeneous types. Which IMO, is the same problem a little less obvious. They will have different methods and not be quite compatible (similar to the integers which are not quite compatible). I feel the main alternative would be to always return datetime64. But then we need some way to convert to `datetime.datetime` more conveniently probably.

**Comment by h-vetinari** [comment]
> @seberg: I feel the main alternative would be to always return datetime64. But then we need some way to convert to `datetime.datetime` more conveniently probably.

This would be the best solution I think. People should explicitly cast to `datetime.datetime` (through whatever mechanism) if they so wish, and deal with the potential fallout themselves.

**Comment by seberg** [comment]
Well, one annoying thing about that is that there is no good way to give a FutureWarning (at least I cannot think of one right now). If you currently use `astype(object)`, and say you actually prefer to get an object array of `datetime64` (why?!) there is no way to get that without a warning. Which is very annoying for libraries.

**Comment by h-vetinari** [comment]
I know it's not an easy change. But the reasons for wanting `datetime64` are many:
* not having ints for results that are outside `datetime.datetime`
* not having unexpected casting with `.astype(object)`
* not losing the performance of `np.datetime64` just because the array might have to contain some non-datetime-data.
* ...

Couldn't the current `array.astype(object)` check if the result contains a `datetime.datetime` and warn in that case? I know that - worst case - this would mean scanning all of the data (types).

**Comment by seberg** [comment]
Giving the warning is easy, this behaviour only happens when the warning is shown. When we add a warning, it is good to also provide a way to avoid the warning. `np.asarray(arr.tolist(), dtype=object)` may be made to work to give the old behaviour, even if a bit annoying. There would be no way to get the new without a warning.

**Comment by h-vetinari** [comment]
> @seberg: There would be no way to get the new without a warning.

The warning could be restricted to once per session, which should be (more or less) an acceptable level of annoyance?

**Comment by seberg** [comment]
For end-users yes, but I feel that libraries should always be able to avoid warnings, since they confuse their end-users. Otherwise they will be tempted to ignore the warning and that is also a can of worms. We could try to argue it is the lesser evil.

There may be another complexity to this whole thing because of gh-15411...

**Comment by gerritholl** [comment]
I regularly want to extract times from `xarray.DataArray` coordinates and then use them in plotting labels.  This is currently cumbersome due to the present casting behaviour:

Given:

```python
import xarray as xr
import numpy as np
da = xr.DataArray([0, 1], dims=("time",), coords={"time": np.array([0, 1], dtype="M8[s]")})
```

* `f'{da.coords["time"][0]:%Y}'` fails with `TypeError: unsupported format string passed to DataArray.__format__` because `datetime64` does not support the `datetime` format specifiers,
* `f'{da.coords["time"][0].item():%Y}'` fails with `ValueError: Invalid format specifier` because xarray time coordinates (like pandas timeseries) have nanosecond precision and therefore the `.item()` call results in an integer,
* f'{da.coords["time"][0].values.astype("M8[s]").item():%Y}' currently works, but would break if `.item()` started to return `datetime64` by default.  I use this approach a lot in my code.
* f'{da.coords["time"][0].values.astype("M8[s]").astype(datetime.datetime):%Y}' would probably have to be the new approach if the default behaviour of `item()` were to change, enless `datetime64` grows format specifiers.

I would be interested what alternative workaround to "format xarray time coordinate" other people may suggest.  

Personally, my preferred alternative would be for:

- `datetime64` to support format specifiers (I don't know if there is an issue for this),
- `item()` to return `datetime64` at all times,
- `.astype(datetime.datetime)` to truncate the precision, just like `float32(2.34).astype(int32)` does today — there should be no expectation that type conversions are lossless.  For out of range I'm less sure (currently, `datetime64("NaT").astype(datetime.datetime)` returns `None`; depending on what's in `np.errstate`, an `OverflowError` may be appropriate).

**Comment by charris** [comment]
Pushing this off again.

**Comment by jbrockmendel** [comment]
I've been finding a lot of subtle pandas bugs caused by this caused by this.  We've mostly collected the appropriate patches https://github.com/pandas-dev/pandas/blob/master/pandas/core/array_algos/putmask.py#L196 with a smaller one https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1696 

Is there someone qualified to address this who I could incentivize to prioritize it?

**Comment by seberg** [comment]
I guess I can try to actually get it done.  @jbrockmendel what would the desired change be, was this approach to your liking?:

1. We will deprecate any return which is not a python datetime
2. It will give an error in the future.
3. If someone has a use-case, we may try to provide a utility function.

**Comment by h-vetinari** [comment]
I'm not involved in pandas-development anymore, but then the problem transcends any one library like pandas. I think that
> We will deprecate any return which is not a python datetime

is the wrong approach - it should [IMO](https://github.com/numpy/numpy/issues/12550#issuecomment-568278922) be "anything but `np.datetime64` gets deprecated". See also the comment from @gerritholl above.

**Comment by jbrockmendel** [comment]
> I guess I can try to actually get it done

Thanks.  LMK if i can reciprocate, e.g pandas bugs youd like to see prioritized.

> We will deprecate any return which is not a python datetime

I'd expect an object array of np.datetime64 objects.  stdlib datetimes have different implementation bounds and resolutions.  Also there is no pydatetime NaT.

**Comment by seberg** [comment]
Well, that is probably even easier to achieve. But it means we change the return value of a whole lot of operations...

My problem here is not the change itself (it really isn't difficult), but making the decision that its worth whatever breakage may happen. Let me make a PR to see what *always* datetie64 would look like (that is even for `.item()`!).

**Comment by h-vetinari** [comment]
I just re-read this thread on this occasion, and regarding migration pains, I think a once-per-session warning is feasible (and I believe library authors will be able to work around it for the transition) - see discussion starting [here](https://github.com/numpy/numpy/issues/12550#issuecomment-567991073).

There was also a [suggestion](https://github.com/numpy/numpy/issues/12550#issuecomment-450577569) by @eric-wieser to consider `.astype(np.generic)`, which would likely open easier migration avenues, but is perhaps a bigger can of worms still because it requires a new (meta-)type, and would require changing peoples expectations / habits.

**Comment by seberg** [comment]
Well, if all you want is to make `.item()` *and* `.astype(object)` return NumPy scalars, below is the diff you need.

If you want a warning for `np.array(...).astype(object)` (or limit to `astype`), I have to wait for my current PR to go in in the next days, it touches too much code to ignore.  Basically, we need to at least make sure a warning happens only once per `asarray` call.

I don't like `np.generic`, but other name could be possible (again after my PR is in).  Also `astype(datetime.datetime)` (maybe with a different name) is plausible (not quite _yet_ maybe). But I feel that should not return an `object` array per-se. 

I _honestly_ don't have much of an opinion here, but I am wary of warnings that cannot be silenced or may not be seen. Maybe datetimes are not used all that much (at least in raw-form), but I just don't know.


<details>

```
diff --git a/numpy/core/src/multiarray/arraytypes.c.src b/numpy/core/src/multiarray/arraytypes.c.src
index ecaca72a1..8ed6e2ee8 100644
--- a/numpy/core/src/multiarray/arraytypes.c.src
+++ b/numpy/core/src/multiarray/arraytypes.c.src
@@ -968,48 +968,14 @@ VOID_setitem(PyObject *op, void *input, void *vap)
 static PyObject *
 DATETIME_getitem(void *ip, void *vap)
 {
-    PyArrayObject *ap = vap;
-    npy_datetime dt;
-    PyArray_DatetimeMetaData *meta = NULL;
-
-    /* Get the datetime units metadata */
-    meta = get_datetime_metadata_from_dtype(PyArray_DESCR(ap));
-    if (meta == NULL) {
-        return NULL;
-    }
-
-    if ((ap == NULL) || PyArray_ISBEHAVED_RO(ap)) {
-        dt = *((npy_datetime *)ip);
-    }
-    else {
-        PyArray_DESCR(ap)->f->copyswap(&dt, ip, PyArray_ISBYTESWAPPED(ap), ap);
-    }
-
-    return convert_datetime_to_pyobject(dt, meta);
+    return PyArray_Scalar(ip, PyArray_DESCR((PyArrayObject *)vap), NULL);
 }
 
 
 static PyObject *
 TIMEDELTA_getitem(void *ip, void *vap)
 {
-    PyArrayObject *ap = vap;
-    npy_timedelta td;
-    PyArray_DatetimeMetaData *meta = NULL;
-
-    /* Get the datetime units metadata */
-    meta = get_datetime_metadata_from_dtype(PyArray_DESCR(ap));
-    if (meta == NULL) {
-        return NULL;
-    }
-
-    if ((ap == NULL) || PyArray_ISBEHAVED_RO(ap)) {
-        td = *((npy_timedelta *)ip);
-    }
-    else {
-        PyArray_DESCR(ap)->f->copyswap(&td, ip, PyArray_ISBYTESWAPPED(ap), ap);
-    }
-
-    return convert_timedelta_to_pyobject(td, meta);
+    return PyArray_Scalar(ip, PyArray_DESCR((PyArrayObject *)vap), NULL);
 }
 
 static int
```

**Comment by h-vetinari** [comment]
> I have to wait for my current PR to go in in the next days, it touches too much code to ignore.

Just referencing here (what I believe to be) the referenced PR: #18398

**Comment by h-vetinari** [comment]
@seberg

Now that #18398 is merged, should this be tackled? Would you like a PR with the diff you proposed?

**Comment by seberg** [comment]
Well, ideally yes. Although, there is some dtype stuff I really want to focus on.

The diff leads to crashes for some scalar operations IIRC.  I think the diff is correct, but probably an infinite recursion has to be broken somewhere (some code in NumPy assumes that most types will return a Python scalar with `arr.item()` and similar. Datetimes (and float128) don't, so they need extra care sometimes.

**Comment by charris** [comment]
@h-vetinari Are you still working on this?

**Comment by h-vetinari** [comment]
> 
> 
> @h-vetinari Are you still working on this?

Hey @charris, thanks for checking in; fundamentally yes, I still want to fix it, but since I read somewhere that 1.21 branches in May, that's not gonna happen until then. #18683 is my first foray into the entrails of numpy, and that's a pretty large initial hurdle to overcome (and work/life hasn't been kind recently).

**Comment by rgommers** [comment]
Bumped to `1.23.0`, doesn't seem blocking for `1.22.0`

**Comment by jbrockmendel** [comment]
Started a branch with the just the diff proposed by @seberg [here](https://github.com/numpy/numpy/issues/12550#issuecomment-780203343), very similar to #18683.  

Eventually ended up scaling it back to only try to change the cases that give integers.  So the diff looks like

```
diff --git a/numpy/core/src/multiarray/arraytypes.c.src b/numpy/core/src/multiarray/arraytypes.c.src
index 71401c60e..733413edd 100644
--- a/numpy/core/src/multiarray/arraytypes.c.src
+++ b/numpy/core/src/multiarray/arraytypes.c.src
@@ -1024,6 +1024,11 @@ DATETIME_getitem(void *ip, void *vap)
         PyArray_DESCR(ap)->f->copyswap(&dt, ip, PyArray_ISBYTESWAPPED(ap), ap);
     }
 
+    // See https://github.com/numpy/numpy/issues/12550
+    if (PyLong_Check(convert_datetime_to_pyobject(dt, meta))) {
+        return PyArray_Scalar(ip, PyArray_DESCR((PyArrayObject *)vap), NULL);
+    }
+
     return convert_datetime_to_pyobject(dt, meta);
 }
 
@@ -1048,6 +1053,11 @@ TIMEDELTA_getitem(void *ip, void *vap)
         PyArray_DESCR(ap)->f->copyswap(&td, ip, PyArray_ISBYTESWAPPED(ap), ap);
     }
 
+    // See https://github.com/numpy/numpy/issues/12550
+    if (PyLong_Check(convert_timedelta_to_pyobject(td, meta))) {
+        return PyArray_Scalar(ip, PyArray_DESCR((PyArrayObject *)vap), NULL);
+    }
+
     return convert_timedelta_to_pyobject(td, meta);
 }
```

and added tests

```
diff --git a/numpy/core/tests/test_datetime.py b/numpy/core/tests/test_datetime.py
index baae77a35..4d44e7b1a 100644
--- a/numpy/core/tests/test_datetime.py
+++ b/numpy/core/tests/test_datetime.py
@@ -1557,6 +1557,37 @@ def test_hours(self):
         t[0] = 60*60*24 + 60*60*10
         assert_(t[0].item().hour == 10)
 
+    def test_astype_object_not_ints(self):
+        dts = np.ones(3, dtype=f'M8[ns]')
+        dts[0] = 60*60*24 + 60*60*10
+        assert isinstance(dts[0].item(), np.datetime64)
+
+        obj_dt = dts.astype(object)
+        assert isinstance(obj_dt[0], np.datetime64)
+        assert obj_dt[0].dtype == dts.dtype  # i.e. same resolution
+        assert isinstance(obj_dt[:1].item(), np.datetime64)
+
+        tds = dts.view(f"m8[ns]")
+        obj_td = tds.astype(object)
+        assert isinstance(obj_td[0], np.timedelta64)
+        assert obj_td[0].dtype == tds.dtype  # i.e. same resolution
+        assert isinstance(obj_td[:1].item(), np.timedelta64)
+
+    @pytest.mark.parametrize("unit", ["us", "ms", "s", "m", "h", "D", "W", "M", "Y"])
+    def test_astype_object_not_ints_oob(self, unit):
+        dts = np.array([11000], dtype="M8[Y]").astype(f"M8[{unit}]")
+
+        res = dts.astype(object)
+        assert isinstance(res[0], np.datetime64)
+        assert res[0].dtype == dts.dtype
+
+        tds = dts - np.datetime64(0, unit)
+        res = tds.astype(object)
+        if not isinstance(res[0], datetime.timedelta):
+            # we just want NOT an integer
+            assert isinstance(res[0], np.timedelta64)
+            assert res[0].dtype == tds.dtype
```

With this, everything in numpy/core/tests/test_datetime.py passes.  But I still get a segfault in numpy/core/tests/test_array_coercion.py::TestTimeScalars::test_coercion_timedelta_convert_to_number

I also tried to just call convert_timedelta_to_pyobject/convert_datetime_to_pyobject once in each function but that gave compile-time errors that I punted on.

@h-vetinari think you can get this over the finish line?

**Comment by h-vetinari** [comment]
> @jbrockmendel: @h-vetinari think you can get this over the finish line?

Started looking at this again. I don't have a compiler stack set up on the only machine I have easily available currently, so I'm mostly restricted to CI. Might set up a VM if it becomes necessary.

> @seberg: some code in NumPy assumes that most types will return a Python scalar with `arr.item()` and similar. Datetimes (and float128) don't, so they need extra care sometimes.

I started digging a bit, but I'm not familiar at all with the numpy code-base. Any pointers which pieces I should look at to avoid the kind of recursion you mentioned?

**Comment by seberg** [comment]
Pushing this off again unfortunately since branching is getting close.  Please ping to bump, although IIRC the whole thing was a bit tricky to fix unfortunately.

**Comment by seberg** [comment]
I don't know if anyone has the cycles/enthusiasm to dive into this one again.  But I fixed some infinite recursions here recently which may make this more feasible.
(Of course it is still a breaking change for things that know how to do deal with Python datetimes but not NumPy ones and currently can work via `datetime_arr + other_obj`.)
